Nightly failures with in builds with gcc/9.1, 9.2 in toolchain: TestBitManipulationBuiltins.hpp: error: no matching function for call to ‘TestBitCastFunction<Space>::check...

**Describe the bug**
Nightly builds with gcc/9.1 or gcc/9.2 in the toolchain failed to compile the `TestBitManipulationBuiltins.hpp` unit test following merge of #6101  with error output:

cuda/11.7 and cuda/11.2 with gcc/9.2
```
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:812:27: error: no matching function for call to ‘TestBitCastFunction<Space>::check<int>(float) const’
09:12:03   812 |       if (!check<int>(12.34f)) {
09:12:03       |                           ^
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:792:1: note: candidate: ‘template<class Space> template<class To, class From> static bool TestBitCastFunction<Space>::check(const From&)’
09:12:03   792 |   static KOKKOS_FUNCTION bool check(const From& from) {
09:12:03       | ^ ~~~
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:792:1: note:   template argument deduction/substitution failed:
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:812:27: note:   mismatched types ‘const From’ and ‘float’
09:12:03   812 |       if (!check<int>(12.34f)) {
09:12:03       |                           ^
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:818:43: error: no matching function for call to ‘TestBitCastFunction<Space>::check<long long unsigned int>(double) const’
09:12:03   818 |       if (!check<unsigned long long>(123.456)) {
09:12:03       |                                           ^
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:792:1: note: candidate: ‘template<class Space> template<class To, class From> static bool TestBitCastFunction<Space>::check(const From&)’
09:12:03   792 |   static KOKKOS_FUNCTION bool check(const From& from) {
09:12:03       | ^ ~~~
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:792:1: note:   template argument deduction/substitution failed:
09:12:03 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkosdev2_cuda117_gcc920_cpp17/kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:818:43: note:   mismatched types ‘const From’ and ‘double’
09:12:03   818 |       if (!check<unsigned long long>(123.456)) {

```

@dalg24 can you investigate?

Reproducer:
```bash
module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/9.2.0 cuda/11.7

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,Serial --arch=Volta70 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="17" --kokkos-path=$KOKKOS_PATH --with-cuda-options=enable_lambda  --no-examples   --with-options=enable_large_mem_tests
```

--- a/core/unit_test/TestBitManipulationBuiltins.hpp
+++ b/core/unit_test/TestBitManipulationBuiltins.hpp
@@ -789,7 +789,7 @@ struct TestBitCastFunction {
     ASSERT_EQ(errors, 0) << "Failed check no error for bit_cast()";
   }
   template <typename To, typename From>
-#if defined(KOKKOS_COMPILER_GNU) && (910 <= KOKKOS_COMPILER_GNU) && \
+#if defined(KOKKOS_COMPILER_GNU) && (900 <= KOKKOS_COMPILER_GNU) && \
     (KOKKOS_COMPILER_GNU < 930)
   // workaround compiler bug in GCC 9.1 and GCC 9.2
   KOKKOS_FUNCTION bool check(const From& from) const

--- a/core/unit_test/TestBitManipulationBuiltins.hpp
+++ b/core/unit_test/TestBitManipulationBuiltins.hpp
@@ -789,7 +789,14 @@ struct TestBitCastFunction {
     ASSERT_EQ(errors, 0) << "Failed check no error for bit_cast()";
   }
   template <typename To, typename From>
-  static KOKKOS_FUNCTION bool check(const From& from) {
+#if defined(KOKKOS_COMPILER_GCC) && (910 <= KOKKOS_COMPILER_GCC) && \
+    (KOKKOS_COMPILER_GCC < 930)
+  // workaround compiler bug in GCC 9.1 and GCC 9.2
+  KOKKOS_FUNCTION bool check(const From& from) const
+#else
+  static KOKKOS_FUNCTION bool check(const From& from)
+#endif
+  {
     using Kokkos::Experimental::bit_cast_builtin;
     return bit_cast_builtin<From>(bit_cast_builtin<To>(from)) == from;
   }

--- a/core/unit_test/TestBitManipulationBuiltins.hpp
+++ b/core/unit_test/TestBitManipulationBuiltins.hpp
@@ -791,7 +791,7 @@ struct TestBitCastFunction {
   template <typename To, typename From>
 #if defined(KOKKOS_COMPILER_GNU) && (900 <= KOKKOS_COMPILER_GNU) && \
     (KOKKOS_COMPILER_GNU < 930)
-  // workaround compiler bug in GCC 9.1 and GCC 9.2
+  // workaround compiler bug seen in GCC 9.0.1 and GCC 9.2.0
   KOKKOS_FUNCTION bool check(const From& from) const
 #else
   static KOKKOS_FUNCTION bool check(const From& from)

--- a/core/unit_test/TestBitManipulationBuiltins.hpp
+++ b/core/unit_test/TestBitManipulationBuiltins.hpp
@@ -789,8 +789,8 @@ struct TestBitCastFunction {
     ASSERT_EQ(errors, 0) << "Failed check no error for bit_cast()";
   }
   template <typename To, typename From>
-#if defined(KOKKOS_COMPILER_GCC) && (910 <= KOKKOS_COMPILER_GCC) && \
-    (KOKKOS_COMPILER_GCC < 930)
+#if defined(KOKKOS_COMPILER_GNU) && (910 <= KOKKOS_COMPILER_GNU) && \
+    (KOKKOS_COMPILER_GNU < 930)
   // workaround compiler bug in GCC 9.1 and GCC 9.2
   KOKKOS_FUNCTION bool check(const From& from) const
 #else

